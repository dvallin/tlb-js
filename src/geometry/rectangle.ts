import { AbstractShape } from "./shape"
import { Vector } from "../spatial"

export class Rectangle extends AbstractShape {

    public static fromBounds(left: number, right: number, top: number, bottom: number): Rectangle {
        return new Rectangle(left, top, right - left + 1, bottom - top + 1)
    }

    public constructor(
        public readonly x: number,
        public readonly y: number,
        public readonly width: number,
        public readonly height: number,
    ) {
        super()
    }

    public get left(): number {
        return this.x
    }

    public get right(): number {
        return this.x + this.width - 1
    }

    public get top(): number {
        return this.y
    }

    public get bottom(): number {
        return this.y + this.height - 1
    }

    public get topLeft(): Vector {
        return new Vector(this.left, this.top)
    }

    public get topRight(): Vector {
        return new Vector(this.right, this.top)
    }

    public get bottomRight(): Vector {
        return new Vector(this.right, this.bottom)
    }

    public get bottomLeft(): Vector {
        return new Vector(this.left, this.bottom)
    }

    public get center(): Vector {
        const cx = (this.right + this.left) / 2
        const cy = (this.bottom + this.top) / 2
        return new Vector(Math.floor(cx), Math.floor(cy))
    }

    public plus(other: Rectangle): Rectangle {
        return Rectangle.fromBounds(
            Math.min(this.left, other.left), Math.max(this.right, other.right),
            Math.min(this.top, other.top), Math.max(this.bottom, other.bottom)
        )
    }

    public bounds(): Rectangle {
        return this
    }

    public containsVector(p: Vector): boolean {
        return p.x >= this.x && p.x < this.x + this.width
            && p.y >= this.y && p.y < this.y + this.height
    }

    public translate(t: Vector): Rectangle {
        return new Rectangle(this.x + t.x, this.y + t.y, this.width, this.height)
    }

    public grow(): Rectangle {
        return new Rectangle(this.x - 1, this.y - 1, this.width + 2, this.height + 2)
    }

    public all(f: (position: Vector) => boolean): boolean {
        for (let i = 0; i < this.height; i++) {
            for (let j = 0; j < this.width; j++) {
                const position = new Vector(j + this.x, i + this.y)
                if (!f(position)) {
                    return false
                }
            }
        }
        return true
    }
}
