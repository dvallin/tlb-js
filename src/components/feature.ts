import { Color } from '../renderer/color'
import { gray, primary } from '../renderer/palettes'
import { TlbWorld } from '../tlb'

import { FeatureComponent } from './feature'
import { PositionComponent } from './position'
import { WorldMap } from '../resources/world-map'
import { Vector } from '../spatial'
import { Entity } from '../ecs/entity'
import { strangeSymbols } from '../symbols'

export type FeatureType = keyof typeof features
export interface FeatureComponent {
  type: FeatureType
}

export interface Feature {
  character: string
  diffuse: Color

  blocking: boolean
  lightBlocking: boolean
  description: string
}

export const features = {
  wall: {
    character: '#',
    diffuse: gray[3],
    blocking: true,
    lightBlocking: true,
    description: 'a wall',
  },
  corridor: {
    character: '.',
    diffuse: gray[0],
    blocking: false,
    lightBlocking: false,
    description: 'a corridor',
  },
  room: {
    character: '.',
    diffuse: primary[1],
    blocking: false,
    lightBlocking: false,
    description: 'floor of a room',
  },
  locker: {
    character: strangeSymbols[16],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description: 'a locker',
  },
  trash: {
    character: strangeSymbols[21],
    diffuse: gray[2],
    blocking: false,
    lightBlocking: false,
    description: 'some trash',
  },
  door: {
    character: strangeSymbols[27],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description: 'a door',
  },
  hub: {
    character: '.',
    diffuse: primary[0],
    blocking: false,
    lightBlocking: false,
    description: 'floor of a hub',
  },
  player: {
    character: '@',
    diffuse: primary[0],
    blocking: true,
    lightBlocking: true,
    description: 'you',
  },
  guard: enemy('g', 'a guard'),
  eliteGuard: eliteEnemy('g', 'a very strong guard'),
}

function enemy(character: string, description: string): Feature {
  return {
    character,
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description,
  }
}

function eliteEnemy(character: string, description: string): Feature {
  return {
    character,
    diffuse: primary[3],
    blocking: true,
    lightBlocking: true,
    description,
  }
}

export function getFeature(world: TlbWorld, entity: number): Feature | undefined {
  const feature = world.getComponent<FeatureComponent>(entity, 'feature')
  if (feature) {
    return features[feature.type]
  }
  return undefined
}

export function createFeature(world: TlbWorld, map: WorldMap, position: Vector, type: FeatureType): Entity {
  const e = map.getTile(position)
  if (e !== undefined) {
    throw new Error(`there is already a tile at ${position.key}`)
  }

  const entity = world
    .createEntity()
    .withComponent<PositionComponent>('position', { position })
    .withComponent<FeatureComponent>('feature', { type }).entity
  map.setTile(position, entity)
  return entity
}
