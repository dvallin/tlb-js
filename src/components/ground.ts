import { FeatureType } from "./feature"

export interface GroundComponent {
    feature: FeatureType
}
