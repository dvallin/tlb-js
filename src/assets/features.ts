import { Feature } from '../components/feature'
import { gray, primary } from '../renderer/palettes'
import { strangeSymbols } from '../symbols'

function enemy(character: string, name: string): Feature {
  return {
    character,
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    name,
  }
}

function eliteEnemy(character: string, name: string): Feature {
  return {
    character,
    diffuse: primary[3],
    blocking: true,
    lightBlocking: true,
    name,
  }
}

const featuresDefinition = {
  wall: {
    character: '#',
    diffuse: gray[3],
    blocking: true,
    lightBlocking: true,
    name: 'wall',
  },
  corridor: {
    character: '.',
    diffuse: gray[0],
    blocking: false,
    lightBlocking: false,
    name: 'corridor',
  },
  room: {
    character: '.',
    diffuse: primary[1],
    blocking: false,
    lightBlocking: false,
    name: 'floor',
  },
  locker: {
    character: strangeSymbols[16],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    name: 'locker',
  },
  trash: {
    character: strangeSymbols[21],
    diffuse: gray[2],
    blocking: false,
    lightBlocking: false,
    name: 'trash',
  },
  door: {
    character: strangeSymbols[27],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    name: 'door',
  },
  hub: {
    character: '.',
    diffuse: primary[0],
    blocking: false,
    lightBlocking: false,
    name: 'floor',
  },
  player: {
    character: '@',
    diffuse: primary[0],
    blocking: true,
    lightBlocking: true,
    name: 'you',
  },
  loot: {
    character: 'l',
    diffuse: primary[1],
    blocking: false,
    lightBlocking: false,
    name: 'some loot',
  },
  table: {
    character: strangeSymbols[3],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: false,
    name: 'a table',
  },
  guard: enemy('g', 'guard'),
  eliteGuard: eliteEnemy('g', 'elite guard'),
}
export type FeatureType = keyof typeof featuresDefinition
export const features: { [key in FeatureType]: Feature } = featuresDefinition

const generatorsDefinition = {
  block: (index: number) => ({
    character: strangeSymbols[[7, 5, 17, 15][index]],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: false,
    name: 'a block of concrete',
  }),
}
export type FeatureGeneratorsType = keyof typeof generatorsDefinition
export const generators: { [key in FeatureGeneratorsType]: (index: number) => Feature } = generatorsDefinition
