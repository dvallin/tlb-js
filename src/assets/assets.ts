import { Asset, Cover } from '../components/asset'
import { Vector } from '../spatial'
import { features, generators } from './features'
import { Feature } from '../components/feature'
import { DialogType } from './dialogs'

const s1x1 = new Vector([1, 1])
const s1x2 = new Vector([1, 2])
const s3x1 = new Vector([3, 1])
const s2x2 = new Vector([2, 2])
function asset(name: string, size: Vector, cover: Cover, dialog: DialogType | undefined, feature: (index: number) => Feature): Asset {
  return { name, size, cover, hasInventory: true, dialog, feature }
}

const assetsDefinition = {
  door: asset('metal door', s3x1, 'full', undefined, () => features['door']),
  locker: asset('a locker', s1x1, 'full', undefined, () => features['locker']),
  trash: asset('some trash', s1x1, 'none', undefined, () => features['trash']),
  loot: asset('a dead body', s1x1, 'none', undefined, () => features['loot']),
  table: asset('a table', s1x2, 'partial', undefined, () => features['table']),
  generator: asset('a generator', s2x2, 'full', undefined, generators['block']),
  elevator: asset('an elevator', s1x1, 'none', 'elevator', generators['elevator']),
}

export type AssetType = keyof typeof assetsDefinition
export const assets: { [key in AssetType]: Asset } = assetsDefinition
