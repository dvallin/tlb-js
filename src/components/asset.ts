import { Vector } from '../spatial'
import { TlbWorld } from '../tlb'
import { Entity } from '../ecs/entity'
import { WorldMap, WorldMapResource } from '../resources/world-map'
import { Shape } from '../geometry/shape'

import { FeatureComponent, createFeature, Feature, FeatureProvider } from './feature'
import { GroundComponent } from './ground'
import { PositionComponent } from './position'
import { Rectangle } from '../geometry/rectangle'
import { TriggersComponent } from './trigger'

import { AssetType, assets } from '../assets/assets'
import { FeatureType } from '../assets/features'
import { Direction } from '../spatial/direction'
import { InventoryComponent } from './items'
import { DialogType } from '../assets/dialogs'
import { DialogComponent } from './dialog'

export interface AssetComponent {
  type: AssetType
}

export interface Asset {
  name: string
  size: Vector
  hasInventory: boolean
  dialog: DialogType | undefined
  feature: (index: number) => Feature
}

export interface PlacedAsset {
  type: AssetType
  tiles: { position: Vector; feature: FeatureType }[]
}

export function shapeOfAsset(type: AssetType, position: Vector, direction: Direction): Shape {
  const asset = assets[type]
  return Rectangle.footprint(position, direction, asset.size)
}

export function createAsset(world: TlbWorld, level: number, position: Vector, direction: Direction, type: AssetType): Entity {
  const shape = shapeOfAsset(type, position, direction)
  return createAssetFromShape(world, level, shape, type)
}

export function createAssetFromShape(world: TlbWorld, level: number, shape: Shape, type: AssetType): Entity {
  const entity = placeAsset(world, type)
  placeAssetParts(world, type, entity, level, shape)
  return entity
}

function placeAsset(world: TlbWorld, type: AssetType): Entity {
  const asset = assets[type]
  const entity = world
    .createEntity()
    .withComponent<AssetComponent>('asset', { type })
    .withComponent<TriggersComponent>('triggers', { name: asset.name, type: 'asset', entities: [] }).entity
  if (asset.hasInventory) {
    world.editEntity(entity).withComponent<InventoryComponent>('inventory', { content: [] })
  }
  if (asset.dialog !== undefined) {
    world.editEntity(entity).withComponent<DialogComponent>('dialog', { type: asset.dialog })
  }
  return entity
}

function placeAssetParts(world: TlbWorld, type: AssetType, entity: Entity, level: number, shape: Shape): void {
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const asset = assets[type]
  shape.foreach((p, i) => {
    const tile = putTile(world, map, level, p, () => asset.feature(i))
    addTrigger(world, tile, entity)
  })
}

export function removeAsset(world: TlbWorld, map: WorldMap, entity: Entity): void {
  const triggers = world.getComponent<TriggersComponent>(entity, 'triggers')!
  triggers.entities.forEach(tile => removeTile(world, map, tile))
  world.deleteEntity(entity)
}

function removeTile(world: TlbWorld, map: WorldMap, entity: Entity): void {
  const ground = world.getComponent<GroundComponent>(entity, 'ground')!
  const position = world.getComponent<PositionComponent>(entity, 'position')!
  map.levels[position.level].removeTile(position.position)
  createFeature(world, position.level, position.position, ground.feature)
  world.deleteEntity(entity)
}

function putTile(world: TlbWorld, map: WorldMap, level: number, position: Vector, feature: FeatureProvider): Entity {
  const ground = removeGround(world, map, level, position)
  const tile = world
    .createEntity()
    .withComponent<PositionComponent>('position', { level, position })
    .withComponent<FeatureComponent>('feature', { feature })
    .withComponent<GroundComponent>('ground', { feature: ground }).entity
  map.levels[level].setTile(position, tile)
  return tile
}

function addTrigger(world: TlbWorld, tile: Entity, entity: Entity): void {
  world.editEntity(tile).withComponent('triggered-by', { entity })
  world.getComponent<TriggersComponent>(entity, 'triggers')!.entities.push(tile)
}

function removeGround(world: TlbWorld, map: WorldMap, level: number, position: Vector): FeatureProvider {
  const entity = map.levels[level].removeTile(position)
  if (entity === undefined) {
    throw new Error(`cannot build asset on missing ground at ${position.key}`)
  }
  const feature = world.getComponent<FeatureComponent>(entity, 'feature')
  if (feature === undefined || feature.feature().blocking) {
    throw new Error('cannot build asset on missing or blocking ground')
  }
  world.deleteEntity(entity)
  return feature.feature
}
