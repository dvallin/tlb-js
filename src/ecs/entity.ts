import { World } from "./world"

export type Entity = number

export class EntityModifier<C, S, R> {

    public constructor(
        private readonly world: World<C, S, R>,
        public readonly entity: Entity
    ) { }

    public withComponent<T>(name: C, component: T): EntityModifier<C, S, R> {
        this.world.getStorage(name).insert(this.entity, component)
        return this
    }

    public removeComponent(name: C): EntityModifier<C, S, R> {
        this.world.getStorage(name).remove(this.entity)
        return this
    }

    public delete(): void {
        this.world.components.forEach(s => s.remove(this.entity))
    }
}
