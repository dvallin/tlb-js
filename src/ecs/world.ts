import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
