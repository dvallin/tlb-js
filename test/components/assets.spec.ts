import { createAssetFromShape, createAsset } from '../../src/components/asset'
import { Vector } from '../../src/spatial'
import { World } from '../../src/ecs/world'
import { TlbWorld, registerComponents, registerResources } from '../../src/tlb'
import { WorldMap, WorldMapResource } from '../../src/resources/world-map'
import { FeatureComponent } from '../../src/components/feature'
import { mockRenderer } from '../mocks'
import { Rectangle } from '../../src/geometry/rectangle'
import { Entity } from '../../src/ecs/entity'
import { features } from '../../src/assets/features'
import { GroundComponent } from '../../src/components/ground'
import { Random } from '../../src/random'
import { Uniform } from '../../src/random/distributions'

describe('createAssetFromPosition', () => {
  let world: TlbWorld
  let map: WorldMap
  let random: Random
  beforeEach(() => {
    jest.clearAllMocks()
    world = new World()
    registerComponents(world)
    registerResources(world, mockRenderer())
    map = world.getResource<WorldMapResource>('map')
    random = new Random(new Uniform('createAssetFromPosition'))
  })

  it('throws error on missing ground', () => {
    // given
    const position = new Vector([0, 0])

    // when / then
    expect(() => createAsset(world, random, 0, position, 'up', 'door')).toThrowErrorMatchingSnapshot()
  })

  it('throws error on blocking ground', () => {
    // given
    const position = new Vector([0, 0])
    const ground = world.createEntity().withComponent<FeatureComponent>('feature', { feature: () => features['wall'] }).entity
    map.levels[0].setTile(position, ground)

    // when / then
    expect(() => createAsset(world, random, 0, position, 'up', 'door')).toThrowErrorMatchingSnapshot()
  })

  it('creates assets', () => {
    // given
    const shape = new Rectangle(0, 0, 2, 2)
    const entities: Entity[] = []
    shape.foreach(position => {
      const ground = world.createEntity().withComponent<FeatureComponent>('feature', { feature: () => features['corridor'] }).entity
      map.levels[0].setTile(position, ground)
      entities.push(ground)
    })

    // when
    const asset = createAssetFromShape(world, random, 0, shape, 'door')

    // then
    shape.foreach(position => {
      const tile = map.levels[0].getTile(position)!
      expect(world.getComponent(tile, 'position')).toEqual({ position, level: 0 })
      expect(world.getComponent<GroundComponent>(tile, 'ground')!.feature()).toEqual(features['corridor'])
      expect(world.getComponent<FeatureComponent>(tile, 'feature')!.feature()).toEqual(features['door'])
      expect(world.getComponent(tile, 'triggered-by')).toEqual({ entity: asset })
    })

    expect(world.getComponent(asset, 'triggers')).toEqual({ entities, type: 'asset', name: 'metal door' })
    expect(world.getComponent(asset, 'asset')).toEqual({ type: 'door' })
  })
})
