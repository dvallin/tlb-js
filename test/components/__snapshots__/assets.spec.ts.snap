// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAssetFromPosition throws error on blocking ground 1`] = `"cannot build asset on missing or blocking ground"`;

exports[`createAssetFromPosition throws error on missing ground 1`] = `"cannot build asset on missing ground at 0,0"`;
